ecs-deploy

```diff
@@ -1,408 +1,467 @@
+            # new version of the task definition
+
+            RUNNING_TASKS=$($AWS_ECS list-tasks --cluster "$CLUSTER"  --service-name "$SERVICE" --desired-status RUNNING \
+                | jq -r '.taskArns[]')
+
+            if [[ ! -z $RUNNING_TASKS ]] ; then
+                RUNNING=$($AWS_ECS describe-tasks --cluster "$CLUSTER" --tasks $RUNNING_TASKS \
+                    | jq ".tasks[]| if .taskDefinitionArn == \"$NEW_TASKDEF\" then . else empty end|.lastStatus" \
+                    | grep -e "RUNNING") || :
 
-      RUNNING_TASKS=$($AWS_ECS list-tasks --cluster "$CLUSTER"  --service-name "$SERVICE" --desired-status RUNNING \
-          | jq -r '.taskArns[]')
+                if [ "$RUNNING" ]; then
+                    echo "Service updated successfully, new task definition running.";
 
-      if [[ ! -z $RUNNING_TASKS ]] ; then
-        RUNNING=$($AWS_ECS describe-tasks --cluster "$CLUSTER" --tasks $RUNNING_TASKS \
-          | jq ".tasks[]| if .taskDefinitionArn == \"$NEW_TASKDEF\" then . else empty end|.lastStatus" \
-          | grep -e "RUNNING") || :
+                    if [[ $MAX_DEFINITIONS -gt 0 ]]; then
+                        FAMILY_PREFIX=${TASK_DEFINITION##*:task-definition/}
+                        FAMILY_PREFIX=${FAMILY_PREFIX%*:[0-9]*}
+                        TASK_REVISIONS=`$AWS_ECS list-task-definitions --family-prefix $FAMILY_PREFIX --status ACTIVE --sort ASC`
 
-        if [ "$RUNNING" ]; then
-          echo "Service updated successfully, new task definition running.";
+                        NUM_ACTIVE_REVISIONS=$(echo "$TASK_REVISIONS" | jq ".taskDefinitionArns|length")
 
-          if [[ $MAX_DEFINITIONS -gt 0 ]]; then
-            FAMILY_PREFIX=${TASK_DEFINITION##*:task-definition/}
-            FAMILY_PREFIX=${FAMILY_PREFIX%*:[0-9]*}
-            TASK_REVISIONS=`$AWS_ECS list-task-definitions --family-prefix $FAMILY_PREFIX --status ACTIVE --sort ASC`
+                        if [[ $NUM_ACTIVE_REVISIONS -gt $MAX_DEFINITIONS ]]; then
+                            LAST_OUTDATED_INDEX=$(($NUM_ACTIVE_REVISIONS - $MAX_DEFINITIONS - 1))
+                            for i in $(seq 0 $LAST_OUTDATED_INDEX); do
+                                OUTDATED_REVISION_ARN=$(echo "$TASK_REVISIONS" | jq -r ".taskDefinitionArns[$i]")
 
-            NUM_ACTIVE_REVISIONS=$(echo "$TASK_REVISIONS" | jq ".taskDefinitionArns|length")
+                                echo "Deregistering outdated task revision: $OUTDATED_REVISION_ARN"
 
-            if [[ $NUM_ACTIVE_REVISIONS -gt $MAX_DEFINITIONS ]]; then
-                LAST_OUTDATED_INDEX=$(($NUM_ACTIVE_REVISIONS - $MAX_DEFINITIONS - 1))
-                for i in $(seq 0 $LAST_OUTDATED_INDEX); do
-                    OUTDATED_REVISION_ARN=$(echo "$TASK_REVISIONS" | jq -r ".taskDefinitionArns[$i]")
+                              $AWS_ECS deregister-task-definition --task-definition "$OUTDATED_REVISION_ARN" > /dev/null
```

Comment:
As for extraction here with the deregister outdated definitions code I wrote a while ago... we could try

```js
function taskRevisionFamilyPrefix() {
    FAMILY_PREFIX=${TASK_DEFINITION##*:task-definition/}
    FAMILY_PREFIX=${FAMILY_PREFIX%*:[0-9]*}

    if [ "$BASH_SOURCE" != "$0" ]; then
        echo $FAMILY_PREFIX
    fi
}
```

And integration test that??? Maybe???? As for testing the later functionality, we would have to use bash function arguments, which is really meh. I could look into writing the tests for this chunk in the future, after the test framework code is merged.